crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1,32 +1,89 @@
+use crate::semantic_index::ast_ids::{HasScopedUseId, ScopedExpressionId};
+use crate::semantic_index::definition::{Definition, DefinitionKind};
+use crate::semantic_index::symbol::{NodeWithScopeRef, ScopeId};
+use crate::semantic_index::usedef::Expression;
+use crate::semantic_index::SemanticIndex;
+use crate::types::{
+    definitions_ty, use_def_map, ClassType, FunctionType, Name, Type, UnionTypeBuilder,
 };
-use crate::semantic_index::{symbol_table, SemanticIndex};
-use crate::types::{infer_types, ClassType, FunctionType, Name, Type, UnionTypeBuilder};
 use crate::Db;
+use ruff_db::parsed::parsed_module;
 
-/// The inferred types for a single scope.
+use crate::semantic_index::semantic_index;
+use crate::semantic_index::symbol::NodeWithScopeKind;
+
+/// Infer all types for a [`Definition`] (including sub-expressions).
+#[salsa::tracked(return_ref)]
+pub(crate) fn infer_definition_types<'db>(
+    db: &'db dyn Db,
+    definition: Definition<'db>,
+) -> TypeInference<'db> {
+    let _span = tracing::trace_span!("infer_definition_types", ?definition).entered();
+
+    let index = semantic_index(db, definition.file(db));
+
+    let mut builder = TypeInferenceBuilder::new(db, InferenceRegion::Definition(definition), index);
+
+    builder.infer_region();
+    builder.finish()
+}
+
+/// Infer all types for an [`Expression`] (including sub-expressions).
+#[allow(unused)]
+#[salsa::tracked(return_ref)]
+pub(crate) fn infer_expression_types<'db>(
+    db: &'db dyn Db,
+    expression: Expression<'db>,
+) -> TypeInference<'db> {
+    let _span = tracing::trace_span!("infer_expression_types", ?expression).entered();
+
+    let index = semantic_index(db, expression.file(db));
+
+    let mut builder = TypeInferenceBuilder::new(db, InferenceRegion::Expression(expression), index);
+
+    builder.infer_region();
+    builder.finish()
+}
+
+/// Infer all types for `scope`.
+#[salsa::tracked(return_ref)]
+pub(crate) fn infer_scope_types<'db>(db: &'db dyn Db, scope: ScopeId<'db>) -> TypeInference<'db> {
+    let _span = tracing::trace_span!("infer_scope_types", ?scope).entered();
+
+    let file = scope.file(db);
+    // Using the index here is fine because the code below depends on the AST anyway.
+    // The isolation of the query is by the return inferred types.
+    let index = semantic_index(db, file);
+
+    let mut builder = TypeInferenceBuilder::new(db, InferenceRegion::Scope(scope), index);
+
+    builder.infer_region();
+    builder.finish()
+}
+
+/// A region within which we can infer types.
+pub(crate) enum InferenceRegion<'db> {
+    Expression(Expression<'db>),
+    Definition(Definition<'db>),
+    Scope(ScopeId<'db>),
+}
+
+/// The inferred types for a single region.
 #[derive(Debug, Eq, PartialEq, Default, Clone)]
 pub(crate) struct TypeInference<'db> {
-    /// The types of every expression in this scope.
+    /// The types of every expression in this region.
     expressions: IndexVec<ScopedExpressionId, Type<'db>>,
```

Comment:
I don't think we can use an `IndexVec` here any longer because each region only stores a very small subset of the expressions. I'm not quiet sure why the tests are passing but I would expect that all expression ids are off.

Reply:
Yes, I think you're right. Tests are passing because at this stage tests are only actually exercising per-definition types; without type narrowing, nothing is currently using the stored per-expression types. I'll either add some unit tests or figure out what feature I can add to this PR that would actually exercise per-expression types in an integration test.